[IMP] project_timesheet_time_control: ❤️ 4 UX

yajo opened the pull request:
This is a huge improvement in this module's usability, although basic functionality stays the same. Let me summmarize:

- In `account.analytic.line` views, the *Stop* button is replaced by colorful *Resume/Stop* buttons.
- The new *Resume* button will display a wizard that will duplicate all meaningful fields from the previous `account.analytic.line` record and auto-stop the previous running timer, if any.
- If the previous running timer is very old, some warnings will be displayed.
- In `project.project` and `project.task` views, I added the same buttons (well... it's *Start* instead of *Resume*, because you could not have worked before in that project or task, but it behaves the same).
- If using it from a `project.task`, *Start* will duplicate meaningful fields from your latest line in that task.
- If using it from a `project.project`, *Start* will duplicate meaningful fields your latest line in that project that had no task associated.
- All kanban, kanban mobile, list and form views are covered.
- I removed, BTW, some Bootstrap 3 and bad replacements in some views.

---

@Tecnativa TT19205
Inspired by https://github.com/OCA/timesheet/pull/139.

Tardo commented on project_timesheet_time_control/models/account_analytic_line.py:
```diff
@@ -3,33 +3,79 @@
     @api.model
     def create(self, vals):
-        return super(AccountAnalyticLine, self).create(self.eval_date(vals))
+        return super(AccountAnalyticLine, self).create(self._eval_date(vals))
```
```suggestion
        return super().create(self._eval_date(vals))
```

Tardo commented on project_timesheet_time_control/models/account_analytic_line.py:
```diff
@@ -3,33 +3,79 @@
     @api.multi
     def write(self, vals):
-        return super(AccountAnalyticLine, self).write(self.eval_date(vals))
+        return super(AccountAnalyticLine, self).write(self._eval_date(vals))
```
```suggestion
        return super().write(self._eval_date(vals))
```

chienandalu reviewed (commented):
These improvements are great!

A little issue:

- Go to timesheet and create two lines succesively.
- The timer of the first one doesn't stop so the times overlap.

Is it intended?

yajo commented:
> A little issue:
> 
>     * Go to timesheet and create two lines succesively.
> 
>     * The timer of the first one doesn't stop so the times overlap.
> 
> 
> Is it intended?

There's no constraint that forbids creating 2 lines with 0 time. Maybe I should add a unique constraint to avoid that?

pedrobaeza commented:
I don't think we should control that.

yajo commented:
That's what I though at first also. After having these controls in place, it's more likely that you create your lines from projects/tasks view than timesheet ones.

:thinking: Maybe I could add a button to create a line from scratch from the menu, opening the wizard without smar fills, to let user autostop latest line.

pedrobaeza commented:
Uhm, maybe we can make a similar screen to the attendance one with big button for starting and stopping.

yajo commented:
That'd be much harder than just adding a menu item that opens the wizard. It would involve JS. I think the menu item should be enough

pedrobaeza commented:
Well, let's put it on the roadmap. I don't see the menu too much added value.

yajo commented:
I thought about it and it makes sense. I added this menu: *Timesheets > Timesheet > Start work*

It opens this wizard, without autocompletes. In the screenshot you can see the obvious benefit of doing it this way instead of just creating a new record: it warns about an old running timer and lets you go to it or stop it and start a new timer with one click:

![Captura de pantalla de 2019-10-01 12-31-38](https://user-images.githubusercontent.com/973709/65958717-f96d1f80-e447-11e9-99c0-9c4a7ca5d7df.png)


So, although I think we will use it less because the new buttons in tasks and projects are so cool, I still think it's good to have it for those that want to input a new line from scratch.

After all, it's just an action and menu item, so easy it's done already.

Tardo reviewed (approved):
LGTM :+1: Awesome improvements!

pedrobaeza commented:
Please add Spanish translation

yajo commented:
Done. I just have to add some tests and try to ensure the wizard always features a hotkey to make it more handy.

yajo commented:
> try to ensure the wizard always features a hotkey

Enforcing a hotkey is very error-prone in a wizard, it's easier to rely on the automatic one, as far as there are available, so I opened https://github.com/OCA/timesheet/pull/278 and patched a little bit kanban and tree views here to add more chances for the wizard button to get a hotkey. Now it will have one almost always.

pedrobaeza commented:
Can you improve a bit more test coverage?

yajo commented:
Yes, it's on my roadmap, do not merge still please.

rafaelbn reviewed (approved):
Awesome! Functionally :+1:

yajo commented:
If I'm not wrong, this should go 100% :umbrella: now.

yajo commented:
Please do not merge yet. I still have to change one thing to make it compatible with the upcoming patch for `crm_timesheet`.

yajo committed:
[IMP] project_timesheet_time_control: ❤️ 4 UX

This is a huge improvement in this module's usability, although basic functionality stays the same. Let me summmarize:

- In `account.analytic.line` views, the *Stop* button is replaced by colorful *Resume/Stop* buttons.
- The new *Resume* button will display a wizard that will duplicate all meaningful fields from the previous `account.analytic.line` record and auto-stop the previous running timer, if any.
- If the previous running timer is very old, some warnings will be displayed.
- In `project.project` and `project.task` views, I added the same buttons (well... it's *Start* instead of *Resume*, because you could not have worked before in that project or task, but it behaves the same).
- If using it from a `project.task`, *Start* will duplicate meaningful fields from your latest line in that task.
- If using it from a `project.project`, *Start* will duplicate meaningful fields your latest line in that project that had no task associated.
- All kanban, kanban mobile, list and form views are covered.
- Access the new wizard directly from a menu item if you prefer.
- I removed, BTW, some Bootstrap 3 and bad replacements in some views.

yajo commented:
Ready to merge.

rafaelbn commented:
@Yajo travis is :red_circle:

yajo commented:
:heavy_check_mark: now.

pedrobaeza commented:
/ocabot merge major

yajo commented:
Once merged let's continue with https://github.com/OCA/timesheet/pull/279.
